### PR TITLE
Add option to set the core affinity for the Timer Svc Task on SMP systems

### PR DESCRIPTION
**Pull request by sudeep-mohanty**
Title
-----------
Make Timer Service Task core affinity configurable

Description
-----------
This PR introduces the `configTIMER_SERVICE_TASK_CORE_AFFINITY` config option which allows a user to set the core affinity of the Timer Service Task on SMP systems. The motivation to have this feature comes from requests such as [this](https://github.com/espressif/esp-idf/issues/12204#issuecomment-1709830113) and [this](https://github.com/espressif/esp-idf/issues/10315) on the ESP-IDF repository. It would be nice if the FreeRTOS-Kernel can support this natively.

Test Steps
-----------
N.A

Checklist:
----------
- [x] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------
None

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timers.c [hunk 1]
```diff
@@ -243,43 +251,89 @@
...
+            if( xTimerQueue != NULL )
+            {
+                #if ( configSUPPORT_STATIC_ALLOCATION == 1 )
+                {
+                    StaticTask_t * pxTimerTaskTCBBuffer = NULL;
+                    StackType_t * pxTimerTaskStackBuffer = NULL;
+                    uint32_t ulTimerTaskStackSize;
+
+                    vApplicationGetTimerTaskMemory( &pxTimerTaskTCBBuffer, &pxTimerTaskStackBuffer, &ulTimerTaskStackSize );
+                    xTimerTaskHandle = xTaskCreateStaticiAffinitySet( prvTimerTask,
```
Comment (archigup) [inline timers.c:305]:
```suggestion
                    xTimerTaskHandle = xTaskCreateStaticAffinitySet( prvTimerTask,
```

**Comment by archigup** [comment]
Hey, thanks for the PR! 

Could the `#if ( configNUMBER_OF_CORES == 1 )` be moved inwards to just around the task create functions? Seems the rest of the logic in the blocks are the same.

**Comment by sudeep-mohanty** [comment]
Thanks for the review @archigup. Updated as suggested.

**Comment by chinglee-iot** [comment]
@sudeep-mohanty 
Thank you for creating the feedback. The issue in your description indicates that timer task is pinned to core 0 and blocked by higher priority tasks on core 0. As a result, the application watchdog callback can not be called within timeout. The alternative solution also suggests that allowing tasks to run on any core could be a better solution, which is the current setting for timer task. Therefore, this should not be an issue in FreeRTOS SMP.

On the other hand, user can consider to create another task with core affinity to serve application specific requirement instead of changing the affinity of timer task cause it may have effect on software timer service.

**Comment by sudeep-mohanty** [comment]
Thanks for the feedback @chinglee-iot. The idea here is to give users the flexibility of either punning the timer task to a core or letting it run unpinned. Ofcourse, all issues with a lower priority task getting blocked by a higher priority task remain even on an SMP system irrespective of the task's core affinity, so IMO this PR shouldn't cause change to how the timer task runs in practice. A counter argument could be to let the timer task run on a core uninterrupted for a potentially more responsive timer service which is not possible today.

**Review by aggarg — APPROVED** [review]

**Review by chinglee-iot — APPROVED** [review]

**Pull request merged**